Design document for taskman arbitrary linux process service

mikewurtz opened the pull request:
Design document for the taskman service, gives high level design of the implementation of the L4 challenge: https://github.com/gravitational/careers/blob/main/challenges/systems/challenge-1.md#level-4.

The taskman service will include three major components
1. gRPC server
2. CLI to communicate with gRPC server
3. Reusable library to implement core functionality.

mikewurtz committed:
initial design document for taskman

tigrato commented on docs/taskman-design.md:
```diff
@@ -0,0 +1,325 @@
+  		Display help information for the start command.
+
+Example:
+  $ taskman start "/bin/ls /myFolder" --user-id client001
```
Is there a way of using the command without quotes?
having to quote the command + args seems a bit annoying 😆

mikewurtz replied:
Sure, we should be able to support both quoted commands and non-quoted. We can maybe model off of how `kubectl exec` works. Require the flags first and then follow up with a -- to signify the command. So the CLI command would look like `taskman start --user-id client001 -- ls /myfolder`

We should also be able to support commands that are quoted `taskman start --user-id client001 -- "ls /myfolder"` or include chaining `taskman start --user-id client001 -- "ls /myfolder | grep foo"`. These will need to be executed under a shell though.

tigrato commented on docs/taskman-design.md:
````diff
@@ -0,0 +1,325 @@
+```
+
+#### stop
+The stop command will stop the provided task id if the task is running and owned by the user. If the task is not running this command will have no effect. If the user does not own the task or the task does not exist a `NOT_FOUND` will be returned.
````
For errors are you using gRPC standard error codes?

mikewurtz replied:
Yes planning on using codes from "google.golang.org/grpc/codes" and building the errors from "google.golang.org/grpc/status".

tigrato commented on docs/taskman-design.md:
```diff
@@ -0,0 +1,325 @@
+
+By leveraging the `syscall` library we can set the [SysProcAttr](https://pkg.go.dev/syscall#SysProcAttr) to have `UseCgroupFD` and `CgroupFD` set. This will put the child process into the provided cgroup file descriptor during the fork phase before the exec call; avoiding any time where the child process executes outside of the cgroup. This requires the use of cgroup v2, go 1.22 or higher, and li ...
+
+When the stop CLI command is executed, the library will first try and cleanly terminate the process with a SIGTERM but if the process is still running after a grace period a SIGKILL will be sent.
```
what will be the grace period?

mikewurtz replied:
Somewhere between 10-30 seconds seems appropriate, but I am open to suggestions. Docker uses a 10-second grace period, while Kubernetes uses a 30-second grace period. 

Also note that this grace period happens asynchronously; when a stop request is made, the user will receive an OK back immediately, indicating that the request was accepted.

mikewurtz replied:
Based on comment: https://github.com/mikewurtz/taskman/pull/1#discussion_r2027234354 I will omit the grace period.

tigrato commented on docs/taskman-design.md:
```diff
@@ -0,0 +1,325 @@
+The actual values may be different in the actual implementation of the library. These values will be hard coded and not configurable for the exercise. The IO bandwidth restriction will be limited to just one block device with device number Major:Minor 8:0.
+
+#### Streaming
+Streaming will be implemented to support multiple clients streaming the output of a task at the same time and each stream call will start at the top of the output of the task. The output stream will combine the stdout and stderr process outputs and capture their output in goroutines. In a production system we would consider allowing the user to specify if they want stdout or stderr streams separa ...
```
Can't you just pass the same io.Writer to Stderr and Stdout instead of combining them in a goroutine?

mikewurtz replied:
Yes, based on the documentation we should be able to combine them 

```
// If Stdout and Stderr are the same writer, and have a type that can
// be compared with ==, at most one goroutine at a time will call Write.
```

tigrato commented on docs/taskman-design.md:
```diff
@@ -0,0 +1,325 @@
+#### Streaming
+Streaming will be implemented to support multiple clients streaming the output of a task at the same time and each stream call will start at the top of the output of the task. The output stream will combine the stdout and stderr process outputs and capture their output in goroutines. In a production system we would consider allowing the user to specify if they want stdout or stderr streams separa ...
+
+When a new stream request is made, the user will "subscribe" to the task's output stream and receive updates through a Go channel. These updates will be pushed to each user through their channel so they receive real-time updates. The stream will remain open until the process terminates. A rate limiter will be used to throttle the frequency of messages sent, ensuring that bursts of output do not o ...
```
Can you describe how this channel system will handle slow consumers?

rosstimothy replied:
> A rate limiter will be used to throttle the frequency of messages sent

Feel free to omit this for simplicity.

tigrato commented on docs/taskman-design.md:
is there any guarantee that the job output is utf8 data?

mikewurtz replied:
Good point! That is not guaranteed and I should use `bytes` here.

tigrato commented on docs/taskman-design.md:
can you make this field optional and populate it only when the program stops?

mikewurtz replied:
Yes, we can make it `optional int32 exit_code = 2;` and the generated struct should have an ExitCode that is *int32 that we can check if it is nil.

mikewurtz replied:
Great suggestion as this would have been 0 before which is a valid exit code.

tigrato commented on docs/taskman-design.md:
Can you transform this bool into an enum where you can carry more information such as:
- job started
- job was stopped
- job terminated normally
- job exited with an error...

mikewurtz replied:
Sure, I will update the proto

rosstimothy commented on docs/taskman-design.md:
```diff
@@ -0,0 +1,325 @@
+
+Arguments:
+  <command>
+        The full command to execute, including the executable path.
```
Is the full path actually required? What would happen if I ran `taskman start --user-id 5 "ls -lah /home/alice"`?

mikewurtz replied:
No, exec should be able to able to find the command assuming they are in the PATH. Will update

rosstimothy commented on docs/taskman-design.md:
````diff
@@ -0,0 +1,325 @@
+```
+
+#### get-status
+The get-status command is how the user can query an existing task by task-id. The task must have been created by the user in order for the server to return the task details. If the task does not exist or is not associated with the user a `NOT_FOUND` will be returned.
````
> The task must have been created by the user in order for the server to return the task details

Is this accurate? Shouldn't the admin be able to retrieve the status for any task?

mikewurtz replied:
You are correct, I will update this and the other spots you highlighted.

rosstimothy commented on docs/taskman-design.md:
````diff
@@ -0,0 +1,325 @@
+```
+
+#### stop
+The stop command will stop the provided task id if the task is running and owned by the user. If the task is not running this command will have no effect. If the user does not own the task or the task does not exist a `NOT_FOUND` will be returned.
````
> If the user does not own the task

Is this accurate? Shouldn't the admin be able to terminate a task?

rosstimothy commented on docs/taskman-design.md:
````diff
@@ -0,0 +1,325 @@
+```
+
+#### stream
+The stream command will provide users with a stream of the output given the provided `task-id`. If the task is running, the user will receive the output as it is produced continuously starting from the top of the output. If the task is not running, the user will receive the output until the end of the output stream and then close. If the task is running and the task completes or is stopped; the s ...
````
> or is not associated with the use

Same question as above, shouldn't the admin be allowed to stream the task output?

rosstimothy commented on docs/taskman-design.md:
````diff
@@ -0,0 +1,325 @@
+* Engineering:  @tigrato || @timothyb89 || @russjones || @zmb3 || @rosstimothy || @r0mant
+```
+
+### UX
````
What about the UX for starting the server?

mikewurtz replied:
Starting the server will be just running the binary created for the server. I will include a flag for allowing a `--server-address` to be included but will pick a sensible default if not provided. A `--help` command is also probably a good idea. I can include other flags that you see fit. 

Will update the document

rosstimothy commented on docs/taskman-design.md:
```diff
@@ -0,0 +1,325 @@
+        Example: "/bin/ls"
+
+Options:
+  --user-id <user-id>
+  		The user or client ID issuing the request (e.g., client001). This flag is required.
+  --help
```
Should there also be a flag for setting the address of the server, with a sensible default?

mikewurtz replied:
Sure, I was going to hard code this for the exercise but this should be pretty simple to include.

rosstimothy commented on docs/taskman-design.md:
```diff
@@ -0,0 +1,325 @@
+
+By leveraging the `syscall` library we can set the [SysProcAttr](https://pkg.go.dev/syscall#SysProcAttr) to have `UseCgroupFD` and `CgroupFD` set. This will put the child process into the provided cgroup file descriptor during the fork phase before the exec call; avoiding any time where the child process executes outside of the cgroup. This requires the use of cgroup v2, go 1.22 or higher, and li ...
+
+When the stop CLI command is executed, the library will first try and cleanly terminate the process with a SIGTERM but if the process is still running after a grace period a SIGKILL will be sent.
```
For simplicity you may omit graceful termination altogether.

rosstimothy commented on docs/taskman-design.md:
What algorithms will be used to generate the certificates?

mikewurtz replied:
Certificates will be generated with openssl and use Public Key Algorithm: rsaEncryption, Public-Key: (4096 bit), Signature Algorithm: sha256WithRSAEncryption. I include this info in the doc

mikewurtz committed:
chore: respond to pull request feedback

mikewurtz committed:
chore: include StreamTaskOutputResponse should use bytes

mikewurtz committed:
chore: minor updates to stop command due to no grace period

mikewurtz committed:
chore: minor typo fix and stop description update

mikewurtz committed:
fix: server should allow server listening address not just port

mikewurtz committed:
fix: minor alignment

rosstimothy commented on docs/taskman-design.md:
```diff
@@ -0,0 +1,379 @@
+#### Streaming
+Streaming will support multiple clients concurrently consuming the output of a task. Each stream will begin at the start of the task's output. The output will include both stdout and stderr, which will be captured by using the same io.Writer for both. This ensures that output is collected in a consistent and thread-safe manner. In a production system, we may allow clients to specify whether they  ...
+
+When a new stream request is made, the user will "subscribe" to the task's output stream and receive updates through a dedicated Go channel. Each channel is buffered with a fixed size to protect against unbounded memory usage. If a client cannot keep up and their buffer reaches the threshold, the server will close the connection to avoid blocking other consumers.
```
Can you think of a way to achieve this without forcefully terminating slow clients?

mikewurtz replied:
Instead of terminating the stream, we could detect when a client's buffer is full and drop the most recent message being sent. In this case, we’d want to clearly document that slow consumers may miss messages. A client would be able to stream the output again when they have a faster connection and see all the messages.

Would this approach be preferable to you?

rosstimothy replied:
No, we want to see that _all_ clients are able to consume _all_ of the output - no matter how slow they may be consuming.

rosstimothy commented on docs/taskman-design.md:
Suggestion: maybe note that a stopped job, which will also have a non-zero exit code will not be marked as `JOB_STATUS_EXITED_ERROR`
```suggestion
    // job exited with a non-zero status and was not stopped
```

rosstimothy commented on docs/taskman-design.md:
Suggestion: reword/rename this since any binary in the PATH should be runnable as well.

mikewurtz replied:
Good idea, missed updating this part when we changed from requiring the full path.

timothyb89 commented on docs/taskman-design.md:
```diff
@@ -0,0 +1,379 @@
+  Start a new task by executing the specified command. The command may be passed as a quoted string (for shell interpretation), or as separate arguments if no shell features are needed. The --user-id flag is required to identify the client initiating the request.
+
+Arguments:
+  <command> [args...]
+        The command to execute, followed by any optional arguments. The command can be passed as:
+        - A quoted string for shell interpretation (e.g., "ls /myFolder | grep foo")
+        - A command with space-separated arguments (e.g., ls /myFolder)
+
+        The binary can be a full path or must exist in the system's PATH.
+        Example: "ls"
```
I think this is still a bit ambiguous, since some rule is needed to decide whether or not to use shell interpretation (and which shell, etc).

I'd argue the simpler and better solution might be to let the user's own shell do the parsing, and only support the 2nd case. If the user wants to run a shell command, they can just use e.g. `sh -c 'command with spaces'`

mikewurtz replied:
I like this approach better, I will update.

timothyb89 commented on docs/taskman-design.md:
Do you think it might be simpler to return a gRPC error on failure?

mikewurtz replied:
Yes that seems like a better approach that follows better gRPC design

mikewurtz committed:
Update docs/taskman-design.md

Co-authored-by: rosstimothy <39066650+rosstimothy@users.noreply.github.com>

timothyb89 commented on docs/taskman-design.md:
To add to @rosstimothy's comment a bit, will this state apply to jobs stopped only by `taskman` via user request, or also to jobs killed via signal? (`kill -9`, OOM kill, etc).

mikewurtz replied:
We could get more granular and allow more information back to the user if we change this status to be `JOB_STATUS_SIGNALED` and include in the status object
```
string termination_signal = 8;         // e.g., "SIGTERM, SIGKILL"
string termination_source = 9; // "user", "system", "oom", etc.
```

Would this approach be preffered?

mikewurtz committed:
chore: respond to feedback

- streaming will allow all users to receive all the output no matter their consumption rate
- added more granularity on JobStatus and added termination_signal and termination_source
- updated StopTaskResponse to follow grpc standards

mikewurtz committed:
fix: pid should be int32

tigrato commented on docs/taskman-design.md:
```diff
@@ -0,0 +1,379 @@
+#### Streaming
+Streaming will support multiple clients concurrently consuming the output of a task. Each stream will begin at the start of the task's output. The output will include both stdout and stderr, which will be captured by using the same io.Writer for both. This ensures that output is collected in a consistent and thread-safe manner. In a production system, we may allow clients to specify whether they  ...
+
+When a new stream request is made, the client subscribes to the task's output and receives updates through a dedicated buffered Go channel. The buffer size is fixed to avoid unbounded memory usage. Each subscriber maintains an independent read position in the task’s output buffer. If a client's channel becomes full, the server temporarily stops sending messages to that client's channel and does n ...
```
Imagine a program that writes one payload every millisecond. The client, however, is too slow to keep up, so its read position lags behind, and the channel eventually fills up. Once the client catches up and consumes all messages from the channel, the program stops writing because there's no new data. From the client's perspective, it appears to be caught up and in a "live" state—until a new write happens - imagine it takes several hours. At that point, the server realizes the client is actually far behind (based on its read position), so it pushes a burst of data to help it catch up. This causes the client to lag again, and the cycle repeats.

Now, if we already maintain the read position for each user, is it necessary to continue writing payloads to the channel and consuming memory until it fills up?

Let’s say the interface is something like `chan []byte`. If you allocate a buffer of 1024 messages per client, and each message is ~1024 bytes, then each client consumes ~1MB of memory. Scale that to many clients or larger messages, and it becomes easy for an attacker to overwhelm the server by inflating payload sizes and client counts.

Is there a better approach where we can leverage each client's read position to avoid sending data into a channel until it's truly needed—*and* without having to wait for a new write to detect laggy clients?

mikewurtz replied:
@tigrato we could set up a goroutine for each stream request that maintains a readIndex of where the client is in the shared task output buffer that is kept for the task. Each goroutine would use grpc's stream.Send(msg). stream.Send() is a blocking call and should block us from sending more data until the client is ready for more.

We can also set up a sync.Cond that wakes up the buffers that are waiting on  more input.

tigrato replied:
That approach sounds way better to me. 

Do you need a goroutine per stream if the gRPC server already spawns a goroutine for you?

mikewurtz replied:
The goroutine will be on the server side not the client side. We'll want a goroutine on the server to manage the sending of the messages, keeping track of the client's read index, etc.

mikewurtz replied:
I clarified in the design document. I can see why that was confusing

mikewurtz committed:
fix: further streaming approach refinements

mikewurtz committed:
chore: slight revision

mikewurtz committed:
fix: further clarity of server goroutine

mikewurtz committed:
fix: clarify its goroutines waiting on new output

rosstimothy commented on docs/taskman-design.md:
```diff
@@ -0,0 +1,379 @@
+#### Streaming
+Streaming will support multiple clients concurrently consuming the output of a task. Each stream will begin at the start of the task's output. The output will include both stdout and stderr, which will be captured by using the same io.Writer for both. This ensures that output is collected in a consistent and thread-safe manner. In a production system, we may allow clients to specify whether they  ...
+
+Each client stream request spawns a dedicated goroutine on the server that maintains its own readIndex into a shared task output buffer. This buffer holds the combined stdout and stderr output for the task. The goroutine uses stream.Send() to deliver messages to the client, relying on gRPC’s built-in backpressure mechanism. Send() blocks if the client is slow or has stopped reading, preventing un ...
```
Isn't the gRPC stream handler running in its own goroutine already? Why does another goroutine need to be launched just to forward data to the stream?

mikewurtz replied:
Yes you are correct, we shouldn't need to spawn another go routine inside the grpc call itself. Each grpc request is already a goroutine like you said.

mikewurtz committed:
fix: remove goroutine reference for streaming
